### DISCUSSION TITLE
Why the custom cookie banner sticks with the page when you click sub-headings?

### DISCUSSION THREAD

**Discussion by syeda-git** [squidfunk/mkdocs-material#6788] [Q&A] [answered]
Hi,
On this [page](https://opensource.hcltechsw.com/digital-experience/CF216/get_started/product_overview/), scroll down to the bottom. Don't click Accept All Cookies button. 
Now click on a sub-heading that appears at the right-side of the page.
Notice that now you can scroll down further after the footer. The entire cookie banner sticks with the page. If you reload the page, it appears fine.

Follow the same process on this [page](https://opensource.hcltechsw.com/digital-experience/CF217/get_started/product_overview/). The issue is not there. The difference between these two pages is the Material for Mkdocs version. The first one was built with 9.5.6 and the later one with 9.1.1.

Did anything change in the cookie code after 9.1.1 that's impacting custom cookie?

I am using cookie script in analytics.html in the partials folder.

**Comment by kamilkrzyskow** [comment]
Hello @syeda-git,
I see that the cookie consent approval box is provided by `onetrust`, so a 3rd party. The box floating box is added via their script and then the `onetrust-style` `<style>` tag is added to provide styling for it.

After you click on a sub-heading the `onetrust-style` is removed from the `<head>`, and without the styling the box just "spills out" and is not floating. So either it's an issue with the `onetrust` 3rd party logic or I believe this is some kind of issue with Instant Loading, but I'm not sure why it would delete it. Looking at the https://squidfunk.github.io/mkdocs-material/changelog/ a lot of changes to instant navigation were made after 9.1.1. 

So try disabling Instant Navigation or perhaps there is a way to hook the instant loading with the `location$` observable, similar to this example https://squidfunk.github.io/mkdocs-material/setup/setting-up-site-analytics/#usage and re-add the `onetrust-style`, but I'm not sure.
Another method, aka quick fix, would be to permanently save the inline styling to a file `onetrust.css` and add it to the site:
https://squidfunk.github.io/mkdocs-material/customization/#additional-css

**Reply by kamilkrzyskow** [reply]
Reading the source I think that perhaps there is a bug here, because before there were some checks preventing removal of some tags. The checks didn't really cover the current case, so just a guess:
https://github.com/squidfunk/mkdocs-material/blob/b7fc7bad5e5d1408de7feb82d85e1fe3a8e9e8aa/src/templates/assets/javascripts/integrations/instant/index.ts#L207-L209
The changes were made last month in [9.5.5](https://squidfunk.github.io/mkdocs-material/changelog/#9.5.5)
So try with 9.5.4 to see if the issue was there.

@squidfunk the 3rd party javascript adds a `<style>` tag to the `<head>` and if you select an internal heading link the `<style>` is removed. Is that a bug or should the user somehow hook the `location$` observable and re-add it again 🤔 ?

**Reply by squidfunk** [reply]
You're right that we changed replacement of tags in the header in 9.5.5. This was necessary, because authors might change tags or templates in between instant navigation pages, and the tags needed to be correctly replaced, which they were not before, as reported. My recommendation would be to study the issues that were solved with this release:

- #6582
- #6565
- #6345
- #6334
- #6275
- #6102

If you need to persist or re-add styles upon instant navigation, the way to do it is to execute your script as part of a subscription on the `document$` observable, similar to how this is done for [data tables](https://squidfunk.github.io/mkdocs-material/reference/data-tables/#sortable-tables):

``` js
document$.subscribe(function() {
  var tables = document.querySelectorAll("article table:not([class])")
  tables.forEach(function(table) {
    new Tablesort(table)
  })
})
```

We are currently planning a re-vamp of our documentation and a guide on how to interface with instant navigation will be part of that. I apologize that our documentation is currently lacking more instructions on that, as I was lacking the time to put that together. This should be taken care off in the next months.

**Comment by syeda-git** [comment]
@squidfunk @kamilkrzyskow It seems that the issue has been there since 9.4.2. I don't see the issue in 9.4.1.
Given this case, would you recommend anything?

**Reply by squidfunk** [reply]
Yes, please read https://github.com/squidfunk/mkdocs-material/discussions/6788#discussioncomment-8498415 to learn how to integrate external scripts with instant navigation. If that does not work, please provide more information what exactly does not work. It might also be a good idea to ask the creator of the external script for help, because we don't know what this scripts does, so we can only guess.

**Reply by squidfunk** [reply]
Additionally, please note that we are a very small team of people trying to maintain this project, and we can only offer limited support for integrating external scripts with Material for MkDocs. Our community experts and contributors do their absolute best to help you and your company for free, but please understand that there is also a limit to the free support that we can offer. After all, this is an Open Source project that is largely provided free of charge.

Your company might also consider [sponsoring the project](https://squidfunk.github.io/mkdocs-material/insiders/). Thank you for understanding.

**Reply by kamilkrzyskow** [reply, marked as answer]
Here is a more targeted example to your needs @syeda-git:
[issues-instant.zip](https://github.com/squidfunk/mkdocs-material/files/14330037/issues-instant.zip)

In the `extra.js` write:
> https://squidfunk.github.io/mkdocs-material/customization/#additional-javascript
```js
let __gStyles;

// Subscribe to the document$ to know that there was a site reload
// Restore the removed style from the global variable
window.document$.subscribe(function() {
    if (!__gStyles) return;
    if (document.head.querySelector("#onetrust-style")) return;
    document.head.insertAdjacentElement("beforeend", __gStyles);
})

// Subscribe to the location$ to know that the location change was invoked
// On first location change get the style (it's not removed yet) and assign it to the global variable
window.location$.subscribe(function() {
    let style = document.head.querySelector("#onetrust-style");
    if (!style) return;
    if (__gStyles) return;
    __gStyles = style;
})
```

The script assumes that the style tag id will be `onetrust-style`, also it assumes that the tag isn't purged and can be assigned to a global variable without a deep copy. I don't know if it's intentional that it's not removed completely from memory 🤔 

List of all observables can be found here:
https://github.com/squidfunk/mkdocs-material/blob/41d5633fdf9bdac2d4c469c80fd76ac2f5191070/src/templates/assets/javascripts/bundle.ts#L304-L318

**Reply by syeda-git** [reply]
Thanks @kamilkrzyskow for the fix. It worked internally. I will post on production and let you know.

**Reply by kamilkrzyskow** [reply]
Hey @syeda-git I saw that the commit: https://github.com/HCL-TECH-SOFTWARE/digital-experience/commit/04184eae3a0ef5cd9c9c636990246f1328686fef
was made. It copies the whole "mimic" part which should be redundant, since on your site the `onetrust` logic does add the style tag, there is no need for mimicry.

I would also appreciate if the mention of this discussion or a direct attribution to me would be included in a comment next to the solution (or the commit message/description this ship has sailed) :v:

**Reply by syeda-git** [reply]
Thanks @kamilkrzyskow 
https://github.com/HCL-TECH-SOFTWARE/digital-experience/commit/04184eae3a0ef5cd9c9c636990246f1328686fef#commitcomment-139317821

**Reply by rachel-rigdon-sp** [reply]
@kamilkrzyskow I just wanted to thank you for this solution. I ran into this error recently and was able to resolve it quickly thanks to your clear and straightforward directions. 🙌🏻
